File: graph-algorithms/Topological-Sort/DFS/topo-sort.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <queue>
#include <stack>
#include <algorithm>
#include <functional>

// DFS-based topological sort implementation
template<typename T>
std::vector<T> topologicalSortDFS(const DirectedGraph<T>& graph) {
    if (graph.getVertexCount() == 0) {
        return {};
    }
    
    std::unordered_map<T, int> state; // 0: unvisited, 1: visiting, 2: visited
    std::vector<T> result;
    bool hasCycle = false;
    
    // Initialize all vertices as unvisited
    for (const auto& vertex : graph.getVertices()) {
        state[vertex] = 0;
    }
    
    // Helper function for DFS
    std::function<void(const T&)> dfsVisit = [&](const T& vertex) {
        if (hasCycle) return;
        
        state[vertex] = 1; // Mark as visiting
        
        // Visit all neighbors
        for (const auto& neighbor : graph.getNeighbors(vertex)) {
            if (state[neighbor] == 0) {
                // Unvisited vertex
                dfsVisit(neighbor);
            } else if (state[neighbor] == 1) {
                // Back edge detected - cycle found
                hasCycle = true;
                return;
            }
        }
        
        state[vertex] = 2; // Mark as visited
        result.push_back(vertex);
    };
    
    // Perform DFS on all unvisited vertices
    for (const auto& vertex : graph.getVertices()) {
        if (state[vertex] == 0) {
            dfsVisit(vertex);
            if (hasCycle) {
                throw std::runtime_error("Graph contains a cycle. Topological sort not possible.");
            }
        }
    }
    
    // Reverse the result to get correct topological order
    std::reverse(result.begin(), result.end());
    return result;
}


// // Example usage
// int main() {
//     DirectedGraph<std::string> courseGraph;
    
//     // Add course prerequisites
//     courseGraph.addEdge("Mathematics", "Algorithms");
//     courseGraph.addEdge("Mathematics", "Data Structures");
//     courseGraph.addEdge("Algorithms", "Machine Learning");
//     courseGraph.addEdge("Data Structures", "Database Systems");
//     courseGraph.addEdge("Data Structures", "Operating Systems");
//     courseGraph.addEdge("Operating Systems", "Distributed Systems");
    
//     std::cout << "Course Prerequisites Graph:" << std::endl;
//     courseGraph.printGraph();
    
//     std::cout << "\n=== DFS-Based Topological Sort ===" << std::endl;
//     try {
//         auto dfsOrder = topologicalSortDFS(courseGraph);
//         std::cout << "Course sequence (DFS):" << std::endl;
//         for (size_t i = 0; i < dfsOrder.size(); ++i) {
//             std::cout << (i + 1) << ". " << dfsOrder[i] << std::endl;
//         }
//     } catch (const std::exception& e) {
//         std::cout << "Error: " << e.what() << std::endl;
//     }
    
    
//     // Test with a cyclic graph
//     std::cout << "\n=== Testing with Cyclic Graph ===" << std::endl;
//     DirectedGraph<int> cyclicGraph;
//     cyclicGraph.addEdge(1, 2);
//     cyclicGraph.addEdge(2, 3);
//     cyclicGraph.addEdge(3, 1); // Creates a cycle
    
//     std::cout << "Cyclic Graph:" << std::endl;
//     cyclicGraph.printGraph();
    
//     std::cout << "\nTrying DFS-based sort on cyclic graph:" << std::endl;
//     try {
//         auto result = topologicalSortDFS(cyclicGraph);
//         std::cout << "Unexpected success!" << std::endl;
//     } catch (const std::exception& e) {
//         std::cout << "Caught expected error: " << e.what() << std::endl;
//     }
    
    
//     return 0;
// }

File: graph-algorithms/Topological-Sort/DFS/topo-sort.py
from collections import deque
from typing import TypeVar, Generic, List, Set
from enum import Enum

T = TypeVar('T')

class NodeState(Enum):
    """Represents the state of a node during DFS traversal."""
    UNVISITED = 0
    VISITING = 1
    VISITED = 2

def topological_sort_dfs(graph: 'DirectedGraph[T]') -> List[T]:
    """
    Perform topological sort using Depth-First Search algorithm.
    
    Args:
        graph: A DirectedGraph instance
        
    Returns:
        A list representing the topological order of vertices
        
    Raises:
        ValueError: If the graph contains a cycle
    """
    if graph.is_empty():
        return []
    
    # Track the state of each vertex during DFS
    state: dict[T, NodeState] = {vertex: NodeState.UNVISITED for vertex in graph.get_vertices()}
    result: List[T] = []
    has_cycle = False
    
    def dfs_visit(vertex: T) -> None:
        nonlocal has_cycle
        if has_cycle:
            return
            
        state[vertex] = NodeState.VISITING
        
        # Visit all neighbors
        for neighbor in graph.get_neighbors(vertex):
            if state[neighbor] == NodeState.UNVISITED:
                dfs_visit(neighbor)
            elif state[neighbor] == NodeState.VISITING:
                # Back edge detected - cycle found
                has_cycle = True
                return
                
        state[vertex] = NodeState.VISITED
        result.append(vertex)
    
    # Perform DFS on all unvisited vertices
    for vertex in graph.get_vertices():
        if state[vertex] == NodeState.UNVISITED:
            dfs_visit(vertex)
            if has_cycle:
                raise ValueError("Graph contains a cycle. Topological sort not possible.")
    
    # Reverse the result to get correct topological order
    result.reverse()
    return result

# # Example usage
# if __name__ == "__main__":
#     # Example 1: Course prerequisites
#     course_graph = DirectedGraph[str]()
    
#     # Add course dependencies
#     course_graph.add_edge("Mathematics", "Algorithms")
#     course_graph.add_edge("Mathematics", "Data Structures")
#     course_graph.add_edge("Algorithms", "Machine Learning")
#     course_graph.add_edge("Data Structures", "Database Systems")
#     course_graph.add_edge("Data Structures", "Operating Systems")
#     course_graph.add_edge("Operating Systems", "Distributed Systems")
    
#     print("Course Prerequisites Graph:")
#     print(course_graph)
#     print()
    
#     print("=== DFS-Based Topological Sort ===")
#     try:
#         dfs_order = topological_sort_dfs(course_graph)
#         print("Course sequence (DFS):")
#         for i, course in enumerate(dfs_order, 1):
#             print(f"{i}. {course}")
#     except ValueError as e:
#         print(f"Error: {e}")
#     print()
    
#     print("=== Kahn's Algorithm ===")
#     try:
#         kahn_order = topological_sort_kahn(course_graph)
#         print("Course sequence (Kahn):")
#         for i, course in enumerate(kahn_order, 1):
#             print(f"{i}. {course}")
#     except ValueError as e:
#         print(f"Error: {e}")
#     print()
    
#     # Example 2: Testing with cyclic graph
#     print("=== Testing with Cyclic Graph ===")
#     cyclic_graph = DirectedGraph[int]()
#     cyclic_graph.add_edge(1, 2)
#     cyclic_graph.add_edge(2, 3)
#     cyclic_graph.add_edge(3, 1)  # Creates a cycle
    
#     print("Cyclic Graph:")
#     print(cyclic_graph)
#     print()
    
#     print("Trying DFS-based sort on cyclic graph:")
#     try:
#         result = topological_sort_dfs(cyclic_graph)
#         print("Unexpected success!")
#     except ValueError as e:
#         print(f"Caught expected error: {e}")
#     print()
    
#     print("Trying Kahn's algorithm on cyclic graph:")
#     try:
#         result = topological_sort_kahn(cyclic_graph)
#         print("Unexpected success!")
#     except ValueError as e:
#         print(f"Caught expected error: {e}")
